invenio_github/receivers.py

```diff
@@ -49,39 +57,102 @@ def run(self, event):
+
+            if existing_release:
+                raise ReleaseAlreadyReceivedError(release=existing_release)
+
+            # Create the Release
+            repo_id = event.payload["repository"]["id"]
+            repo_name = event.payload["repository"]["name"]
             try:
-                release = Release.create(event)
-                db.session.commit()
+                repo = Repository.get(repo_id, repo_name)
+            except NoResultFound:
+                raise RepositoryNotFoundError(repo_name)
 
-                # FIXME: If we want to skip the processing, we should do it
-                # here (eg. We're in the middle of a migration).
-                # if current_app.config['GITHUB_PROCESS_RELEASES']:
-                process_release.delay(
-                    release.release_id,
-                    verify_sender=self.verify_sender
+            if repo.enabled:
+                release_object = Release(
+                    release_id=release_id,
+                    tag=event.payload["release"]["tag_name"],
+                    repository=repo,
+                    event=event,
+                    status=ReleaseStatus.RECEIVED,
                 )
-            except (ReleaseAlreadyReceivedError, RepositoryDisabledError) as e:
-                event.response_code = 409
-                event.response = dict(message=str(e), status=409)
-            except RepositoryAccessError as e:
-                event.response = 403
-                event.response = dict(message=str(e), status=403)
+                db.session.add(release_object)
+                return release_object
+            else:
+                raise RepositoryDisabledError(repo=repo)
+
+        try:
+            event_release_id = event.payload["release"]["id"]
+            if event_release_id in state:
+                raise ReleaseAlreadyReceivedError()
+
+            # Lock event release to avoid concurrent processing
+            state.update({event_release_id: event})
+
+            # Create a release
+            # runs db.session.add(release)
+            release = _create_release(event)
+
+            # Process the release
+            async_mode = current_app.config.get("GITHUB_ASYNC_MODE", True)
+            if async_mode:
+                # Since 'process_release' is executed asynchronously, we commit the current state of session
+                db.session.commit()
+                process_release.delay(release.release_id)
+            else:
+                release_api = current_github.release_api_class(release)
+                release_api.process_release()
+
+            # Unlock the event release
+            del state[event_release_id]
+        except (
+            ReleaseAlreadyReceivedError,
+            RepositoryDisabledError,
+            SQLAlchemyError,
+        ) as e:
+            # SQLAlchemyError is risen when two or more events arrive at the same time. The one(s) that lose the race condition will try to commit to db and fail.
```

Comment:
This was encountered before implementing a "state" in the receiver. This "state" act as a lock for a release. E.g. if an event already was received for that release, we do not process the event anymore.

So probably the catch for this generic `SQLAlchemyError` can be removed, however, has to be tested.

I also checked legacy Zenodo and saw some events that failed due to this error.

Reply:
I agree on removing it, at the moment it serves as catch all for errors we might not be aware about.